### PR TITLE
Add MetaTransactionWalletDeployer

### PR DESCRIPTION
**Pull request by bowd**
### Description

This is a factory contract that is responsible for deploying and initialising MetaTransactionWalletProxy contracts and setting up the implementation.
It also keeps track of deployed instances in a mapping so that each external account will have at most one wallet deployed.

### Other changes

I've added this line ...
```solidity
_transferOwnership(msg.sender);
```
... to the MetaTransactionWallet itself at the beginning of the `initialize` function because when the contract is set as an implementation to a proxy the `_owner` is `0x0` initially as it's not set in the constructor and thus the `initialize` call always fails because `setSigner` is `onlyOwner`. 
This way when we're calling `initialize` on the proxy, the owner is temporarily set to the `msg.sender` in order to be able to run the rest of the setup and then ownership gets transferred to the contract itself. 

### Tested

Added tests for all flows.

### Related issues

- Fixes #[issue number here]

### Backwards compatibility

New functionality (and a fix that's backwards compatible)

### CODE REVIEW

**Review by asaj — COMMENTED** [review]
What is the advantage of a factory contract over just deploying things "manually"?

**Comment by bowd** [comment]
> What is the advantage of a factory contract over just deploying things "manually"?

It's cheaper and it also makes more sense as a place to keep the mapping between external accounts and wallets which makes it easily recoverable for Valora as well.

@diminator - I know you brought this up initially, am I missing some other advantages?

**Comment by diminator** [comment]
Not much besides. Was mostly a matter of having efficient gas in feeless onboarding.

**Comment by bowd** [comment]
@asaj There's one more reason I totally spaced on yesterday (tired):
We wanted the API to behave in a predictable manner for clients in the sense that all API calls that result in chain interaction will return a TX hash back to the client within a reasonably small amount of time and then the client will wait for that TX to complete at which point it can move to the next step in the flow.
If we wanted to do things manually it means that the call to deploy would need to make two transactions:
1. Deploy the proxy
2. Set implementation and initialise with the signer

So we had three options here: 
1. Wait for the contract to be deployed and return to the client with the TX hash of the 2nd operation - not really good because the API call could stay open at the worst 5s
2. Split the deploy action into two API calls (deploy and initialise) which are both called by the client
3. Use the factory contract which would result in the deploy operation being a single transaction and would maintain the invariant that any API request that touches the chain results in a single TX.

Going with (3) seamed appropriate because it also gave us the efficiency win and provide a good source of truth for the EOA-to-Wallet mapping.

File: packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+  event WalletDeployed(address indexed owner, address indexed wallet, address implementation);
+  event DeployerStatusGranted(address indexed addr);
+  event DeployerStatusRevoked(address indexed addr);
+
+  /**
+     * @dev Verifies that the sender is allowed to deploy a wallet
+     */
+  modifier onlyCanDeploy() {
+    require(msg.sender == owner() || canDeploy[msg.sender] == true, "not-allowed");
```
Comment (ewilz) [inline packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol]: I'm unfamiliar with any conventions clabs has on revert messages, but if I were debugging I might appreciate something more specific like: "sender not authorized to deploy wallet". "not-allowed", if used elsewhere too, could make debugging difficult?

File: packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    }
+  }
+
+  /**
+     * @notice Used to deploy a MetaTransactionWalletProxy, set the implementation,
+     * initialize, transfer ownership and emit an event.
+     * @param owner The external account which will act as signer and owner of the proxy
+     * @param implementation The address of the implementation which the proxy will point to
+     */
+  function deploy(address owner, address implementation, bytes calldata initCallData)
```
Comment (ewilz) [inline packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol:86]: we should probably add `initCallData` to the docstrings above!

File: packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      _changeDeployerAllowance(initialDeployers[i], true);
+    }
+  }
+
+  /**
+     * @notice Change the deployer status of an address
+     * @param target The address to be allowed as a deployer
+     * @param allowedToDeploy toggle whether the address is allowed or not
+     */
+  function changeDeployerAllowance(address target, bool allowedToDeploy) external onlyOwner {
```
Comment (ewilz) [inline packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol]: `allowance` sounds like an allotted sum of money, especially given this is a monetary app. I feel I never hear the word outside that context. `Permissions` sounds clearer to me. But this is a nitpick and not a blocker if other people are OK with `allowance`

**Comment by bowd** [comment]
@emaG3m Thanks for the feedback! You're right that naming was a bit weird. I also handled the rest of the comments.

File: packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  /**
+     * @notice Implementation of permission change
+     * @param target The address to be allowed as a deployer
+     * @param allowedToDeploy toggle whether the address is allowed or not
+     */
+  function _changeDeployerPermission(address target, bool allowedToDeploy) internal {
+    canDeploy[target] = allowedToDeploy;
+    if (allowedToDeploy == true) {
```
Comment (codyborn) [inline packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol]:
```suggestion
    if (allowedToDeploy) {
```

File: packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  event WalletDeployed(address indexed owner, address indexed wallet, address implementation);
+  event DeployerStatusGranted(address indexed addr);
+  event DeployerStatusRevoked(address indexed addr);
+
+  /**
+     * @dev Verifies that the sender is allowed to deploy a wallet
+     */
+  modifier onlyCanDeploy() {
+    require(
+      msg.sender == owner() || canDeploy[msg.sender] == true,
```
Comment (codyborn) [inline packages/protocol/contracts/common/MetaTransactionWalletDeployer.sol]:
```suggestion
      msg.sender == owner() || canDeploy[msg.sender],
```

File: packages/protocol/test/common/metatransactionwalletdeployer.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        })
+      })
+    })
+
+    describe('when permission is granted', async () => {
+      beforeEach(async () => {
+        changePermissionRes = await deployer.changeDeployerPermission(otherAllowedDeployer, true)
+      })
+
+      it('should have added the allownace and emit an event', async () => {
```
Comment (codyborn) [inline packages/protocol/test/common/metatransactionwalletdeployer.ts]:
```suggestion
      it('should have added the allowance and emit an event', async () => {
```

**Review by codyborn — APPROVED** [review]
LGTM

**Pull request merged**